src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs

```diff
@@ -123,10 +117,51 @@ public int Run(ParseResult parseResult)
             return hasFailed ? ExitCodes.GenericFailure : ExitCodes.Success;
         }
 
-        private void WaitOnMSBuildHandlerPipeConnectionLoop()
+        private async Task<bool> RunMSBuild(ParseResult parseResult, bool allowBinLog)
         {
-            _cancellationToken.Cancel();
-            _namedPipeConnectionLoop.Wait((int)TimeSpan.FromSeconds(30).TotalMilliseconds);
+            int msbuildExitCode;
+
+            if (parseResult.HasOption(TestingPlatformOptions.ProjectOption))
+            {
+                string filePath = parseResult.GetValue(TestingPlatformOptions.ProjectOption);
+
+                if (!File.Exists(filePath))
+                {
+                    VSTestTrace.SafeWriteTrace(() => string.Format(LocalizableStrings.CmdNonExistentProjectFilePathDescription, filePath));
+                    return false;
+                }
+
+                msbuildExitCode = await _msBuildConnectionHandler.RunWithMSBuild(filePath, allowBinLog);
+            }
+            else
+            {
+                // If no filter was provided neither the project using --project,
+                // MSBuild will get the test project paths in the current directory
+                msbuildExitCode = await _msBuildConnectionHandler.RunWithMSBuild(allowBinLog);
+            }
+
+            if (msbuildExitCode != ExitCodes.Success)
+            {
+                VSTestTrace.SafeWriteTrace(() => string.Format(LocalizableStrings.CmdMSBuildProjectsPropertiesErrorMessage, msbuildExitCode));
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool IsBinLogEnabled(List<string> args)
+        {
+            var binLog = args.FirstOrDefault(arg => arg.StartsWith("-bl", StringComparison.OrdinalIgnoreCase));
```

Comment:
Is there no helper we can reuse for that? I have just checked and `dotnet build /bl` does produce binlog so this check is not enough it seems.